Avoid auto-granting special permissions to unarchived apps

chenxiaolong opened the pull request:
Avoid calling `setNewlyInstalledInUserId()` when restoring an archived app to ensure that the special permission auto-granting logic does not reset the user's previous permission state.

Fixes: GrapheneOS/os-issue-tracker#5256

---

Tested:
- All 4 combinations of INERNET on/off (install prompt) and OTHER_SENSORS on/off ("Allow Sensors permission to apps by default" global setting) when installing a new app from the Play Store
- All 8 combinations of INTERNET on/off, OTHER_SENSORS on/off, and POST_NOTIFICATIONS on/off when archiving and restoring an app
  * (POST_NOTIFICATIONS to test a non-`SpecialRuntimePermUtils` permission)

chenxiaolong committed:
Avoid auto-granting special permissions to unarchived apps

Avoid calling setNewlyInstalledInUserId() when restoring an archived app
to ensure that the special permission auto-granting logic does not reset
the user's previous permission state.

Fixes: GrapheneOS/os-issue-tracker#5256

Signed-off-by: Andrew Gunnerson <accounts+github@chiller3.com>

muhomorr2 commented:
IIRC permission state is initialized for all users when app gets installed. It's definitely true for apps that were installed before Android 15 update. 

`isNewlyInstalledInUserId` is GrapheneOS-specific, it's a bug that it returns true during unarchival. Need to add a conditional to `setNewlyInstalledInUserId` in InstallPackageHelper.

chenxiaolong commented:
Thanks, I'll look into that approach instead.

chenxiaolong commented:
Done. Updated to use @muhomorr2's suggested approach.

quh4gko8 commented:
This also needs to be verified and tested that it doesn't call `setNewlyInstalledInUserId` on app updates like the current behavior, not only skipping it on archived apps.

quh4gko8 commented on services/core/java/com/android/server/pm/InstallPackageHelper.java:
```diff
@@ -2623,7 +2623,15 @@ private void updateSettingsInternalLI(AndroidPackage pkg,
+                        permissionParamsBuilder =
+                        new PermissionManagerServiceInternal.PackageInstalledParams.Builder();
+
                 if (userId != UserHandle.USER_ALL) {
```
This will regress on app updates and will mistakenly treat the special runtime permission handling as newly installed app.

quh4gko8 commented on services/core/java/com/android/server/pm/InstallPackageHelper.java:
```diff
@@ -2634,6 +2642,10 @@ private void updateSettingsInternalLI(AndroidPackage pkg,
                     // The caller explicitly specified INSTALL_ALL_USERS flag.
                     // Thus, updating the settings to install the app for all users.
                     for (int currentUserId : allUsers) {
```
Same issue as above.

chenxiaolong commented:
Thanks, fixed in dc4fe44d05504f4dd9148010208a597dc109d51a. That was something I overlooked when moving the `setNewlyInstalledInUserId` calls around. Instead of doing that, I changed it to just store a list of user IDs where the app is archived and then act on it later, where the existing `setNewlyInstalledInUserId` calls are.

inthewaves commented on services/core/java/com/android/server/pm/InstallPackageHelper.java:
```diff
@@ -2623,7 +2623,13 @@ private void updateSettingsInternalLI(AndroidPackage pkg,
                     }
                 }
 
+                final Set<Integer> archivedInUserIds = new ArraySet<>();
```
nit: Maybe a `SparseBooleanArray` would work better here in terms of memory usage. `ArraySet` is generic, so it seems it would auto-box the userIds. 

As a reference, the call to the `setNewlyInstalledInUserId` method later puts the `userId` inside of `mNewlyInstalledInUserIds`, which is a `SparseBooleanArray`, and then uses it like a set inclusion check e.g.

https://github.com/GrapheneOS/platform_frameworks_base/blob/3bc0d7c8ad6972cf72b05dfbb32999a5c9d44a0b/services/core/java/com/android/server/pm/permission/PermissionManagerServiceInternal.java#L454-L456

https://github.com/GrapheneOS/platform_frameworks_base/blob/3bc0d7c8ad6972cf72b05dfbb32999a5c9d44a0b/services/core/java/com/android/server/pm/permission/PermissionManagerServiceInternal.java#L391-L393

chenxiaolong replied:
Sure, I can update that. I originally picked `ArraySet` because AOSP already used it for `previousUserIds` below.

chenxiaolong replied:
Done!
